engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java

```diff
@@ -191,6 +199,51 @@ Text toDisplayString(boolean allowSideEffects, @CachedLibrary("this") InteropLib
     }
   }
 
+  @ExportMessage
+  int identityHashCode(
+      @CachedLibrary(limit = "5") InteropLibrary fieldsInterop,
+      @Cached(value = "computeIdentityHashCode(this, fieldsInterop)", allowUncached = true) int cachedHashCode
```

Comment:
How is the caching supposed to work? `this` is not a constant. I am afraid you cannot cache the hash code like this. I believe we will need to have a `hashCode` field in `Atom` and store the once computed hashCode there to avoid it to mutate.

Reply:
What about caching `this` and adding a guard, like in https://github.com/enso-org/enso/pull/3878/files#diff-449f839629893f92453bb6e8ad4c1cbd365bb493223dcee7db1ce97c3796c1b9R29 ? Or maybe, based on your proposal, every `Atom` can have a `hashCode` compilation final field, and maybe an assumption `isHashCodeComputed`.

Reply:
Resolving, because this is very outdated.